revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,246 @@
+// certificate chain, as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/revocation/result"
+	coreX509 "github.com/notaryproject/notation-core-go/x509"
+	"golang.org/x/crypto/ocsp"
+)
+
+// Options specifies values that are needed to check OCSP revocation
+type Options struct {
+	CertChain   []*x509.Certificate
+	SigningTime time.Time
+	HTTPClient  *http.Client
+}
+
+const (
+	pkixNoCheckOID string = "1.3.6.1.5.5.7.48.1.5"
+	// Max size determined from https://www.ibm.com/docs/en/sva/9.0.6?topic=stanza-ocsp-max-response-size.
+	// Typical size is ~4 KB
+	ocspMaxResponseSize int64 = 20480 //bytes
+)
+
+// CheckStatus checks OCSP based on the passed options and returns an array of
+// result.CertRevocationResult objects that contains the results and error. The
+// length of this array will always be equal to the length of the certificate
+// chain.
+func CheckStatus(opts Options) ([]*result.CertRevocationResult, error) {
+	certResults := make([]*result.CertRevocationResult, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return nil, result.InvalidChainError{Err: errors.New("chain does not contain any certificates")}
+	}
+
+	// Validate cert chain structure
+	if err := coreX509.ValidateCodeSigningCertChain(opts.CertChain, &opts.SigningTime); err != nil {
+		return nil, result.InvalidChainError{Err: err}
+	}
+
+	// Check status for each cert in cert chain
+	var wg sync.WaitGroup
+	for i, cert := range opts.CertChain[:len(opts.CertChain)-1] {
+		wg.Add(1)
+		// Assume cert chain is accurate and next cert in chain is the issuer
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			certResults[i] = certCheckStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	certResults[len(opts.CertChain)-1] = &result.CertRevocationResult{
+		Result: result.ResultNonRevokable,
+		ServerResults: []*result.ServerResult{{
+			Result: result.ResultNonRevokable,
+			Error:  nil,
+		}},
+	}
+
+	wg.Wait()
+	return certResults, nil
+}
+
+func certCheckStatus(cert, issuer *x509.Certificate, opts Options) *result.CertRevocationResult {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return &result.CertRevocationResult{
+			Result:        result.ResultNonRevokable,
+			ServerResults: []*result.ServerResult{errToServerResult("", NoOCSPServerError{})},
+		}
+	}
+
+	serverResults := make([]*result.ServerResult, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		serverResult := checkStatusFromServer(cert, issuer, server, opts)
+		if serverResult.Result == result.ResultOK || serverResult.Result == result.ResultRevoked || errors.Is(serverResult.Error, UnknownStatusError{}) {
```

Comment:
We should check `Error` first before parsing `Result`

Reply:
The purpose of that expression is because a result of ResultUnknown can be caused by a variety of errors. However, the only one that is a direct result of a response with the ocsp.Unknown status is the UnknownStatusError. Does that help clear things up?
I can change it to `(serverResult.Result == result.ResultUnknown && errors.Is(serverResult.Error, UnknownStatusError{}))` for clarity in order to convey this better, I just removed the first part since it is guaranteed for that type of error.

Reply:
Yes, please use the complete form of `(serverResult.Result == result.ResultUnknown && errors.Is(serverResult.Error, UnknownStatusError{}))` for readability and less chance to have bugs.